virtcontainers/acrn_arch_base.go

```diff
@@ -0,0 +1,762 @@
+	// Path is the guest kernel path on the host filesystem.
+	Path string
+
+	// InitrdPath is the guest initrd path on the host filesystem.
+	InitrdPath string
+
+	// Params is the kernel parameters string.
+	Params string
+}
+
+// Config is the acrn configuration structure.
+// It allows for passing custom settings and parameters to the acrn-dm API.
+type Config struct {
+
+	// Path is the acrn binary path.
+	Path string
+
+	// Path is the acrn binary path.
+	CtlPath string
+
+	// Name is the acrn guest name
+	Name string
+
+	// UUID is the acrn process UUID.
+	UUID string
+
+	// Devices is a list of devices for acrn to create and drive.
+	Devices []Device
+
+	// Kernel is the guest kernel configuration.
+	Kernel Kernel
+
+	// Memory is the guest memory configuration.
+	Memory Memory
+
+	acrnParams []string
+
+	// ACPI virtualization support
+	ACPIVirt bool
+
+	// NumCPU is the number of CPUs for guest
+	NumCPU uint32
+}
+
+// MaxAcrnVCPUs returns the maximum number of vCPUs supported
+func MaxAcrnVCPUs() uint32 {
+	return uint32(8)
+}
+
+func newAcrnArch(config HypervisorConfig) acrnArch {
+
+	a := &acrnArchBase{
+		path:                 acrnPath,
+		ctlpath:              acrnctlPath,
+		kernelParamsNonDebug: acrnKernelParamsNonDebug,
+		kernelParamsDebug:    acrnKernelParamsDebug,
+		kernelParams:         acrnKernelParams,
+	}
+
+	a.handleImagePath(config)
+	return a
+}
+
+func (a *acrnArchBase) acrnPath() (string, error) {
+	p := a.path
+	return p, nil
+}
+
+func (a *acrnArchBase) acrnctlPath() (string, error) {
+	ctlpath := a.ctlpath
+	return ctlpath, nil
+}
+
+func (a *acrnArchBase) kernelParameters(debug bool) []Param {
+	params := a.kernelParams
+
+	if debug {
+		params = append(params, a.kernelParamsDebug...)
+	} else {
+		params = append(params, a.kernelParamsNonDebug...)
+	}
+
+	return params
+}
+
+func (a *acrnArchBase) memoryTopology(memoryMb uint64) Memory {
+
+	mem := fmt.Sprintf("%dM", memoryMb)
+	memory := Memory{
+		Size: mem,
+	}
+
+	return memory
+}
+
+func (a *acrnArchBase) capabilities() types.Capabilities {
+	var caps types.Capabilities
+
+	// For devicemapper disable support for filesystem sharing
+
+	caps.SetFsSharingUnsupported()
+	caps.SetBlockDeviceSupport()
+	caps.SetBlockDeviceHotplugSupport()
+
+	return caps
+}
+
+// Valid returns true if the CharDevice structure is valid and complete.
+func (cdev ConsoleDevice) Valid() bool {
+
+	if cdev.Backend != "tty" && cdev.Backend != "pty" &&
+		cdev.Backend != "console" && cdev.Backend != "socket" &&
+		cdev.Backend != "file" {
+		return false
+	}
+
+	return true
+}
+
+// AcrnParams returns the acrn parameters built out of this console device.
+func (cdev ConsoleDevice) AcrnParams(slot int, config *Config) []string {
+
+	var acrnParams []string
+	var deviceParams []string
+
+	acrnParams = append(acrnParams, "-s")
+	deviceParams = append(deviceParams, fmt.Sprintf("%d,virtio-console,", slot))
+
+	if cdev.PortType == ConsoleBE {
+		deviceParams = append(deviceParams, "@")
```

Comment:
What's this for?

Reply:
Preceding with @, marks the port as a console port, otherwise it is a normal virtio serial port. This is used for console.sock.